File: Day09/Day09.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

std::vector <int> getData(std::string fileName);
int findWeakPoint(std::vector<int> data, int preambleLength);

int main(){

    std::vector <int> data = getData("/home/pi/Programming/AdventOfCode/2020/Day09/input.txt");
    int response = findWeakPoint(data,25);
    std::cout << "Answer: " << response << "\n";
}

std::vector <int> getData(std::string fileName){
    std::fstream myfile;
    std::vector<int> data{};
    myfile.open(fileName, std::ios::in);
    if(!myfile){
        std::cout << "Cannot find file: " << fileName << "\n";
        return data;
    }
    std::string line;
    while(getline(myfile, line)){
        data.push_back(std::stoi(line));
    }
    std::cout << "File loaded: " << fileName << "\n";
    myfile.close();
    return data;
}

bool cantFindMatch(std::vector <int> data, int target){
    for(int i = 0; i < data.size(); i++){
        std::vector <int> sub_data = data;
        int val = sub_data[i];
        sub_data.erase(sub_data.begin()+i);
        for(int j=0; j<sub_data.size(); j++){
            if (val +sub_data[j] == target)
                return false;
        }  
    }
    return true;
}

int findWeakPoint(std::vector<int> data, int preambleLength){
    int i = preambleLength;
    while( i <data.size()){
        int target_value = data[i];
        std::vector<int> sub_list{};
        // sub_list = data[i-preamble_length:i]
        for(int j=i-preambleLength; j<i; j++){
            sub_list.push_back(data[j]);
        }
        // for (auto i: sub_list)
        //     std::cout << i << ' ';
        // std::cout<< "\n";
        if (cantFindMatch(sub_list,target_value))
            return data[i];
        i++;
    }
    return -1;
}

File: Day09/Day09.py

def cannot_find_match(data, target):
    for i in range(len(data)):
        sub_data = data[:]
        val = sub_data.pop(i)
        for j in range(len(sub_data)):
            if val + sub_data[j] == target:
                return False
    return True

def find_weak_point(data, preamble_length):
    i = preamble_length
    while i < len(data):
        sub_list = data[i-preamble_length:i]
        target_value = data[i]
        
        if cannot_find_match(sub_list, target_value):
            return i, target_value
        i += 1
    return None, None


def find_contiguous_list(data, target):
    i,j = 0,1
    not_found = True
    while not_found:
        if sum(data[i:j])< target:
            j += 1
        elif sum(data[i:j])> target:
            i += 1
        else:
            return min(data[i:j]) +max(data[i:j])
        if j >= len(data):
            return None


if __name__=='__main__':
    data = []
    with open ('/home/pi/Programming/AdventOfCode/2020/Day09/input.txt') as f:
        data = f.read().split('\n')
        data = [int(d) for d in data]

    # print(data)
    index, target = find_weak_point(data,25)
    print(find_contiguous_list(data,target))